# Landscape Genomics Pipeline - Version 2 
These suit of scripts were developed based on LanGen_pipeline from @rojaff/LanGen_pipeline

These scripts perform the following analysis:

<b>1.</b> SNPs dataset filtering using PCadapt, sNMF, and TESS3 to outlier detection.

<b>2.</b> Assessment of population genetic structure using sNMF, DAPC, TESS3, and PCA.

<b>3.</b> Creating maps for ancestry coefficients from step 2.

<b>4.</b> Estimate Tajima's D, genetic diversity and genetic distance inter and intra populations.

<b>5.</b> Assessment of fine-scale spatial genetic structure using LOESS and Spatial Principal Components (sPCA).

<b>6.</b> Isolation by Distance (IBD) using Mantel Tests and Procrustes Analysis.

<b>7.</b> Selection of environmental data

<b>8.</b> Isolation by Resistance (IBR) using Partial Mantel Tests and distance-based Redundancy Analyses (dbRDA).

<b>9.</b> Identification of outlier (adaptive) loci using PCadapt.

<b>10.</b> Identification of genotype-environment associations (GEA) using Redundancy Analyses (RDA) & Mahalanobis distances.

<b>11.</b> Identification of genotype-environment associations (GEA) using Latent Factor Mixed Models (LFMM2).

<b>12.</b> Plotting detections with a Venn Diagram.

<b>13.</b> Visual assessment of genotype-environment associations using Redundancy Analyses (RDA) on the candidate loci.

<b>14.</b> Mapping adaptive genetic variation using Spatial Principal Components (sPCA).

<b>15.</b> Predicting adaptive genotypes from RDA models and environmental data.
    
